File: java/src/main/java/leetcode/LetterCombinationsofaPhoneNumber.java
package leetcode;

import java.util.LinkedList;
import java.util.List;

/**
 * Given a string containing digits from 2-9 inclusive, return all possible
 * letter combinations that the number could represent.
 *
 * A mapping of digit to letters (just like on the telephone buttons) is given
 * below. Note that 1 does not map to any letters.
 *
 * Example:
 *
 * Input: "23" Output: ["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"].
 * Note:
 *
 * Although the above answer is in lexicographical order, your answer could be
 * in any order you want.
 */
final class LetterCombinationsofaPhoneNumber {

    private LetterCombinationsofaPhoneNumber() {
        // prevent instantiation
    }

    /**
     * Return all possible letter combinations that the digits could represent.
     *
     * @param digits the String of digits
     * @return the List<String> of all possible letter combinations
     */
    public static List<String> letterCombinations(final String digits) {

        LinkedList<String> returnList = new LinkedList<String>();

        if (digits.isEmpty()) {
            return returnList;
        }

        String[] mapping = new String[] {"0", "1", "abc", "def", "ghi", "jkl",
                "mno", "pqrs", "tuv", "wxyz"};
        returnList.add("");
        for (int i = 0; i < digits.length(); i++) {
            int index = Character.getNumericValue(digits.charAt(i));
            while (returnList.peek().length() == i) {
                String top = returnList.remove();
                for (char ch : mapping[index].toCharArray()) {
                    returnList.add(top + ch);
                }
            }
        }

        return returnList;
    }
}

File: python/src/letter_combinations_of_a_phone_number.py
from typing import List


class Solution:
    """
    Given a string containing digits from 2-9 inclusive, return all possible letter combinations that the number could
    represent. Return the answer in any order.

    A mapping of digits to letters (just like on the telephone buttons) is given below. Note that 1 does not map to any
    letters.
    """

    def __init__(self):
        """
        Initialise the Solution.
        """
        self.res = []
        self.digits_to_chars = {
            "2": "abc",
            "3": "def",
            "4": "ghi",
            "5": "jlk",
            "6": "mno",
            "7": "pqrs",
            "8": "tuv",
            "9": "wxyz"
        }

    def letterCombinations(self, digits: str) -> List[str]:
        """
        Given a string containing digits from 2-9 inclusive, return all possible letter combinations that the number
        could represent.

        The time complexity is O(4^N) where N is the length of digits, as there are at most 4 choices per digit.
        Faster than 98.29% of solutions.

        The space complexity is O(N*4^N) as each string is of length N. Less memory than 79.23% of solutions.

        :param digits: A string containing digits from 2-9 inclusive.
        :type digits: Str.
        :return: all possible letter combinations that the number could represent.
        :rtype: Str.
        """
        if digits:
            self.dfs(0, digits, "")
        return self.res

    def dfs(self, i: int, digits: str, combination: str):
        """
        A helper function to perform a Depth-First Search (DFS).

        :param i: The current index.
        :type i: Int.
        :param digits: A string containing digits from 2-9 inclusive.
        :type digits: Str.
        :param combination: The current candidate combination.
        :type combination: Str.
        :return: NoneType.
        :rtype: NoneType.
        """
        # base case
        if len(combination) == len(digits):
            self.res.append(combination)
            return

        for c in self.digits_to_chars[digits[i]]:
            self.dfs(i + 1, digits, combination + c)
